vsock/dialer.go

```diff
@@ -0,0 +1,207 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package vsock
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/log"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type VsockTimeout struct {
+	DialTimeout       time.Duration
+	RetryTimeout      time.Duration
+	RetryInterval     time.Duration
+	ConnectMsgTimeout time.Duration
+	AckMsgTimeout     time.Duration
+}
+
+func DefaultTimeouts() VsockTimeout {
+	return VsockTimeout{
+		DialTimeout:       100 * time.Millisecond,
+		RetryTimeout:      20 * time.Second,
+		RetryInterval:     100 * time.Millisecond,
+		ConnectMsgTimeout: 100 * time.Millisecond,
+		AckMsgTimeout:     1 * time.Second,
+	}
+}
+
+// Dial attempts to connect to the Firecracker host-side vsock at the provided unix
+// path and port. It will retry connect attempts if a temporary error is encountered (up
+// to a fixed timeout) or the provided request is canceled.
+func Dial(ctx context.Context, udsPath string, port uint32) (net.Conn, error) {
+	d := vsockDialer{ctx: ctx, timeout: DefaultTimeouts()}
+	return d.Dial(udsPath, port)
+}
+
+type vsockDialer struct {
+	timeout VsockTimeout
+	ctx     context.Context
```

Comment:
Generally, it's an anti-pattern to store a context in a struct because it is possible to construct and use an object in different contexts (e.g. a smaller scoped context with a shorter timeout). I know it technically doesn't matter for this because only vsock.Dial is exposed, but it's good to keep convention for easier refactoring.

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx

https://pkg.go.dev/context

Reply:
Good point. Dial() could take a context. It is different from other Dial functions do, but these Dial functions are implemented before context.

Reply:
Removed this change and pass the context for dial via input parameter. Listener is another story because we implement the net.Listener interface. At the very least this change doesn't make the problem worse.